Functions (#1487)
hadley/adv-r: Functions.Rmd

`[`
```

> They have typeof `builtin` or `special`:

Reviewer: How about: They have either type `builtin` or type `special`

In this section, you'll learn the formal rules of scoping as well as some of its more subtle details. A deeper understanding of scoping will help you to use more advanced functional programming tools, and eventually, even to write tools that translate R code into other languages.

> R uses __lexical scoping__[^dyn-scope]: it looks up the values of names based on how a function is defined, not how it is called. "Lexical" here is not the English adjective relating to words or a vocabulary. It's a technical CS term that tells us that the scoping rules use a parse-time, rather than a run-time structure. 

Reviewer: ```suggestion
R uses __lexical scoping__[^dyn-scope]: it looks up the values of names based on how a function is defined, not how it is called. "Lexical" here is not the English adjective that means relating to words or a vocabulary. It's a technical CS term that tells us that the scoping rules use a parse-time, rather than a run-time structure. 
```

args(sample)
```
 

> It looks like both `x` and `size` are required, but in fact if the latter is not supplied, `sample()` uses `missing()` to provide a default. If I were to rewrite sample, I'd use an explicit `NULL` to indicate that `size` is not required but can be supplied:

Reviewer: ```suggestion
It looks like both `x` and `size` are required, but if `size` is not supplied, `sample()` uses `missing()` to provide a default. If I were to rewrite sample, I'd use an explicit `NULL` to indicate that `size` is not required but can be supplied:
```

\index{ellipsis|see {...}}
\index{dot-dot-dot|see {...}}

> Functions can have a special argument `...` (pronounced dot-dot-dot). With it, a function can take any number of additional arguments. In other programming languages, this type of argument is often called a _varargs_ (variable arguments), and a function that uses it is said to be variadic. 

Reviewer: ```suggestion
Functions can have a special argument `...` (pronounced dot-dot-dot). With it, a function can take any number of additional arguments. In other programming languages, this type of argument is often called a __varargs__ (short for variable arguments), and a function that uses it is said to be variadic. 
```

`for`(i, 1:10, print(i))
```

> Suprisingly, in R, `for` can be called like a regular function! The same is true for basically every operation in R, which means that knowing the function name of a non-prefix function allows you to override its behaviour. For example, if you're ever feeling particularly evil, run the following code while a friend is away from his or her computer. It will introduce a fun bug: 10% of the time, it will add 1 to any numeric calculation inside the parentheses.

Reviewer: Per other discussion, please revert to their

\index{infix functions} 
\indexc{\%\%}

> Infix functions get their name from the fact the function name comes between its arguments, and hence have two arguments. R comes with a number of built-in infix operators: `:`, `::`, `:::`, `$`, `@`, `^`, `*`, `/`, `+`, `-`, `>`, `>=`, `<`, `<=`, `==`, `!=`, `!`, `&`, `&&`, `|`, `||`, `~`, `<-`, and `<<-`. You can also create your own infix functions that start and end with `%`. Base R uses this pattern to define `%%`, `%*%`, `%/%`, `%in%`, `%o%`, and `%x%`.

Reviewer: Can you please revert? This suggestion loses the connection to "in"

"new " %+% "string"
```

> The names of infix functions are more flexible than regular R functions: they can contain any sequence of characters except the percent symbol (%). You will need to escape any special characters in the string used to define the function, but not when you call it:

Reviewer: ```suggestion
The names of infix functions are more flexible than regular R functions: they can contain any sequence of characters except for `%`. You will need to escape any special characters in the string used to define the function, but not when you call it:
```

x
```

> I say they act like they modify their arguments in place, because, as explained in Section \@ref(modify-in-place) about modify-in-place behaviour, they actually create a modified copy. We can see that by using `tracemem()`:

Reviewer: ```suggestion
I say they act like they modify their arguments in place, because, as explained in Section \@ref(modify-in-place), they actually create a modified copy. We can see that by using `tracemem()`:
```

names(x)
```

is translated into:

```{r}
`*tmp*` <- x
x <- `names<-`(`*tmp*`, `[<-`(names(`*tmp*`), 2, "two"))
rm(`*tmp*`)
```

> (Yes, it really does create a local variable named \*tmp\*, which is removed afterwards.)

Reviewer: ```suggestion
(Yes, it really does create a local variable named `*tmp*`, which is removed afterwards.)
```
